src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Display,
+    str::FromStr,
+};
+
+use cosmian_crypto_core::{bytes_ser_de::Serializable, CsRng};
+use rand::SeedableRng;
+use tiny_keccak::{Hasher, Kmac};
+
+use crate::{
+    core::{
+        EncryptedTable, FindexCallbacks, FindexSearch, FindexUpsert, IndexedValue, KeyingMaterial,
+        Keyword, Uid, UpsertData,
+    },
+    error::FindexErr,
+    interfaces::{
+        generic_parameters::{
+            DemScheme, KmacKey, BLOCK_LENGTH, DEM_KEY_LENGTH, KMAC_KEY_LENGTH, KWI_LENGTH,
+            MASTER_KEY_LENGTH, TABLE_WIDTH, UID_LENGTH,
+        },
+        ser_de::serialize_set,
+    },
+};
+
+pub(crate) struct FindexCloud {
+    pub(crate) token: Token,
+    pub(crate) base_url: Option<String>,
+}
+
+pub(crate) struct Token {
+    index_id: String,
+
+    pub(crate) findex_master_key: KeyingMaterial<MASTER_KEY_LENGTH>,
+
+    fetch_entries_key: Option<[u8; 16]>,
+    fetch_chains_key: Option<[u8; 16]>,
+    upsert_entries_key: Option<[u8; 16]>,
+    insert_chains_key: Option<[u8; 16]>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
+
+        if let Some(fetch_entries_key) = self.fetch_entries_key {
+            keys.push(0);
+            keys.extend(fetch_entries_key);
+        }
+
+        if let Some(fetch_chains_key) = self.fetch_chains_key {
+            keys.push(1);
+            keys.extend(fetch_chains_key);
+        }
+
+        if let Some(upsert_entries_key) = self.upsert_entries_key {
+            keys.push(2);
+            keys.extend(upsert_entries_key);
+        }
+
+        if let Some(insert_chains_key) = self.insert_chains_key {
+            keys.push(3);
+            keys.extend(insert_chains_key);
+        }
+
+        write!(f, "{}{}", self.index_id, base64::encode(keys))
+    }
+}
+
+impl FromStr for Token {
+    type Err = FindexErr;
+
+    fn from_str(token: &str) -> Result<Self, Self::Err> {
+        let (index_id, tail) = token.split_at(5);
```

Comment:
It's the format chosen by me :-) The index ID is 5 characters long (to avoid too much collision but not take a lot of space inside the end-user token).

Reply:
Then it needs to be clearly specified, using a constant and some doc.

Reply:
Yes

Reply:
Is it better for you inside ff89a2a?